Fix Crash on invalid time_step value

SGSSGene opened the pull request:
This PR consist of two commits:
1. Small cleanup and replacing C-Casts with dynamic_casts.
2. Fixing crashes on invalid time_step_size. The issue is when sending invalid time_step via the client the player controller crashes.
  Although the time_step is invalid the sensors gets activated and inserted into the `new_sensors` list, which causes division
  by 0 errors later in the code
This PR fixes this.

medrimonia commented on projects/samples/contests/robocup/controllers/player/player.cpp:
```diff
@@ -392,51 +426,21 @@ class PlayerServer {
-        if (sensor_time_step != 0 && sensor_time_step < basic_time_step)
+        if (sensor_time_step == 0)
+          sensors.erase(device);
+        else if (sensor_time_step != 0 && sensor_time_step < basic_time_step)
```
With the way code is written, we can just remove the `sensor_time_step != 0 &&` right?

SGSSGene replied:
Yes, totally. I fixed this!

medrimonia commented:
Thanks for the contribution after taking a closer look, I noticed that the current version of `release` has been already changed quite significantly by the following commit: af72d51a70477f134a3040e55573501350cb5e6d

This solved an issue where device where enabled multiple times if the same time_step was sent: https://github.com/cyberbotics/webots/issues/3020

I think that it still has the version in `experimental` (that should be merge on Monday) has the same problem you raised here and that it should be saved in a similar manner but a simple merge is not possible there.

SGSSGene committed:
Fix using static_cast for casting dynamic objects

Replaces C-Cast with their proper dynamic_cast calls.

SGSSGene committed:
Fix crashing on invalid time_step_size

If time_step_size was not set to a multiple of basic_time_step the
player controller would crash. The sensor got accidentally activated
although the time_step_size was not a multiple, causing
device->getSamplingPeriod() to return 0, which causes a division by zero
bug.

SGSSGene commented:
> Thanks for the contribution after taking a closer look, I noticed that the current version of `release` has been already changed quite significantly by the following commit: [af72d51](https://github.com/RoboCup-Humanoid-TC/webots/commit/af72d51a70477f134a3040e55573501350cb5e6d)
> 
> This solved an issue where device where enabled multiple times if the same time_step was sent: [cyberbotics#3020](https://github.com/cyberbotics/webots/issues/3020)
> 
> I think that it still has the version in `experimental` (that should be merge on Monday) has the same problem you raised here and that it should be saved in a similar manner but a simple merge is not possible there.

I took a look. Yes it still has the same bug. I rebased my PR onto the experimental branch.

omichel commented:
Note that we should use `static_cast` instead of `dynamic_cast` as we know for sure the class of the object the class test is unnecessary.

medrimonia commented:
@SGSSGene I agree with Olivier's comment, dynamic cast is more expensive than static cast, so since the class of the object is known, I am not sure why we should be using dynamic_cast, have you experienced an `invalid_cast` case?

SGSSGene commented:
Yes you are totally right!  `static_cast` suffices completely.
